plugins/UM3NetworkPrinting/resources/qml/GenericPopUp.qml

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2018 Ultimaker B.V.
+// Cura is released under the terms of the LGPLv3 or higher.
+
+import QtQuick 2.2
+import QtQuick.Controls 2.0
+import QtQuick.Controls.Styles 1.4
+import QtQuick.Dialogs 1.1
+import QtGraphicalEffects 1.0
+import UM 1.3 as UM
+
+/**
+ * This is a generic pop-up element which can be supplied with a target and a content item. The
+ * content item will appear to the left, right, above, or below the target depending on the value of
+ * the direction property
+ */
+Popup
+{
+    id: base
+
+    /**
+     * The target item is what the pop-up is "tied" to, usually a button
+     */
+    property var target
+
+    /**
+     * Which direction should the pop-up "point"?
+     * Possible values include:
+     *   - "up"
+     *   - "down"
+     *   - "left"
+     *   - "right"
+     */
+    property string direction: "down"
+
+    /**
+     * We save the default direction so that if a pop-up was flipped but later has space (i.e. it
+     * moved), we can unflip it back to the default direction.
+     */
+    property string originalDirection: ""
+
+    /**
+     * Should the popup close when you click outside it? For example, this is
+     * disabled by the InfoBlurb component since it's opened and closed using mouse
+     * hovers, not clicks.
+     */
+    property bool closeOnClick: true
+
+    /**
+     * Use white for context menus, dark grey for info blurbs!
+     */
+    property var color: "#ffffff" // TODO: Theme!
+
+    Component.onCompleted:
+    {
+        recalculatePosition()
+
+        // Set the direction here so it's only set once and never mutated
+        originalDirection = (' ' + direction).slice(1)
+    }
+
+    background: Item
+    {
+        anchors.fill: parent
+
+        DropShadow
+        {
+            anchors.fill: pointedRectangle
+            color: UM.Theme.getColor("monitor_shadow")
+            radius: UM.Theme.getSize("monitor_shadow_radius").width
+            source: pointedRectangle
+            transparentBorder: true
+            verticalOffset: 2 * screenScaleFactor
+        }
+
+        Item
+        {
+            id: pointedRectangle
+            anchors
+            {
+                horizontalCenter: parent.horizontalCenter
+                verticalCenter: parent.verticalCenter
+            }
+            height: parent.height - 10 * screenScaleFactor // Because of the shadow
+            width: parent.width - 10 * screenScaleFactor // Because of the shadow
```

Comment:
Isn't there a variable or constant somewhere which defines the size of the shadow?

Reply:
One would wish so, but no. I Googled far and wide and it's only possible if you have a reference to the actual Qt.Window which we do not.